File: abc170/d.rb
require 'prime'

def solve n
	i = 1
	while n * i <= @a_max
		@dp[n * i] = 1
		i += 1
	end
end

n = gets.to_i
as = gets.chomp.split.map(&:to_i).sort
@a_max = as[-1]
@dp = Array.new(@a_max+1,0)
ans = 0

n.times do |i|
	a = as[i]
	ans += 1 if !(i < n - 1 && a == as[i+1]) && @dp[a] == 0
	solve(a) if @dp[a] == 0
end

puts ans

File: abc170/d.cpp
#include <bits/stdc++.h>
using namespace std;

bool solve(int n){
	bool flag = true;
	for(int i=0 ;i<=sqrt(n) ;i++){
		if(n % i == 0 && (m[i] == 1 || m[n/i] == 1){
			flag = false;
			break;
		}
	}
	return flag;
}

int main() {
	map<int,int> m;
	int n,ans=0;
	vector<int> as;
	cin >> n;
	as.resize(n);
  for (int i = 0; i < n; i++)
	{
		cin >> as[i];
	}
	as.sort();
	for (int i = 0; i < n - 1; i++)
	{
		if(as[i] == as[i+1]){
			m[as[i]] = 1;
		}
	}

	for (int i = 0; i < n; i++)
	{
		if(m[as[i]] == 1){
			continue;
		}
		bool flag = solve(as[i]);
		if(flag){
			ans++;
			m[as[i]] = 1;
		}
	}
	cout << ans << endl;
}